### PR TITLE
L2 forwarding

### LINKED ISSUE #84: Implement L2 forwarding
Currently, we have a basic implementation for L2 forwarding.  Stratum has an implementation that we may be able to leverage.

https://github.com/stratum/stratum/blob/master/stratum/pipelines/main/main.p4#L454

### LINKED ISSUE #128: Complete TCP support
Our current TCP tests do not check for ACK packets which currently cannot be received by the sender as the data_forward tables are not being populated for southbound communications.

### LINKED ISSUE #138: Allow southbound traffic on the switches
At present, the packets only travel north up to the inet host and the AE. Once we enable southbound communication, it would help TCP packets to flow through, particularly for the handshake process (SYN, SYN-ACK, ACK).

### PR DESCRIPTION
**Pull request by ghost**
#### What does this PR do?
Fixes #84
Fixes #138
Fixes #128
Removes static table entries to forward IPv4 packets and uses the learning model of an L2 switch to build and write P4 table entries to forward packets.
#### Do you have any concerns with this PR?
The remote tests do not work with all the devices in the same subnet.
#### How can the reviewer verify this PR?
Run CI
#### Questions:
- Have you connected this PR to the issue it resolves? Yes
- Does the documentation need an update? No
- Does this add new dependencies? No
- Have you added unit or functional tests for this PR?

### CODE REVIEW

File: trans_sec/controller/aggregate_controller.py [hunk 1]
```diff
@@ -27,6 +26,7 @@ def __init__(self, platform, p4_build_out, topo, log_dir, load_p4=True):
             platform, p4_build_out, topo, 'aggregate',
             ['TpsAggIngress.forwardedPackets'],
             log_dir, load_p4, 'TpsAggIngress')
+        self.known_devices = set()
```
Comment (spisarski) [inline trans_sec/controller/aggregate_controller.py:29]: please define self.known_devices in AbstractController

File: trans_sec/controller/ddos_sdn_controller.py [hunk 2]
```diff
@@ -63,6 +63,7 @@ def start(self):
         logger.info('Starting Controllers')
         for controller in self.controllers.values():
             controller.start()
+            controller.switch_forwarding()
```
Comment (spisarski) [inline trans_sec/controller/ddos_sdn_controller.py:66]: Add this call into abstract_controller#start() after line 61 after self.__add_helpers.

File: trans_sec/controller/core_controller.py [hunk 3]
```diff
@@ -34,6 +34,7 @@ def __init__(self, platform, p4_build_out, topo, log_dir, load_p4=True):
             platform, p4_build_out, topo, 'core', list(), log_dir, load_p4,
             'TpsCoreIngress')
         self.p4_egress = 'TpsCoreEgress'
+        self.known_devices = set()
```
Comment (spisarski) [inline trans_sec/controller/core_controller.py:37]: please define self.known_devices in AbstractController

File: trans_sec/controller/gateway_controller.py [hunk 4]
```diff
@@ -29,6 +27,7 @@ def __init__(self, platform, p4_build_out, topo, log_dir, load_p4=True):
             platform, p4_build_out, topo, 'gateway',
             ['TpsGwIngress.forwardedPackets', 'TpsGwIngress.droppedPackets'],
             log_dir, load_p4, 'TpsGwIngress')
+        self.known_devices = set()
```
Comment (spisarski) [inline trans_sec/controller/gateway_controller.py:30]: please define self.known_devices in AbstractController

File: trans_sec/controller/gateway_controller.py [hunk 5]
```diff
@@ -95,55 +94,74 @@ def make_rules(self, sw, sw_info, north_facing_links, south_facing_links):
...
+            logger.error(sw_info.get('name'))
+
+    def set_multicast_group(self, sw, sw_info):
+        mc_group_id = 1
+
+        # TODO/FIXME - Need to add logic to parse the topology to determine
+        #     egress ports being used on the switch.
+
+        if sw_info['name'] == 'gateway1':
+            replicas = [{'egress_port': '1', 'instance': '1'},
```
Comment (spisarski) [inline trans_sec/controller/gateway_controller.py:122]: Please add ticket so we don't forget to fix this

File: trans_sec/controller/abstract_controller.py [hunk 6]
```diff
@@ -354,3 +391,9 @@ def __setup_tofino_helper(self, name, switch):
         self.switches.append(new_switch)
         logger.info('Instantiated connection to Tofino switch - [%s]',
                     self.switch_type, name)
+
+    def set_multicast_group(self, switch, sw_info):
+        pass
```
Comment (spisarski) [inline trans_sec/controller/abstract_controller.py:396]: Please raise Unimplemented exception here

File: trans_sec/controller/abstract_controller.py [hunk 7]
```diff
@@ -354,3 +391,9 @@ def __setup_tofino_helper(self, name, switch):
...
+
+    def add_data_forward(self, sw, sw_info, source_ip, source_mac, ingress_port):
+        pass
```
Comment (spisarski) [inline trans_sec/controller/abstract_controller.py:399]: same as above

File: trans_sec/controller/aggregate_controller.py [hunk 8]
```diff
@@ -91,3 +76,55 @@ def add_data_inspection(self, sw, device, sw_info):
...
             ' with MAC - [%s] with action params - [%s]',
             self.switch_type, device.get('mac'), action_params)
+
+    def set_multicast_group(self, sw, sw_info):
+        mc_group_id = 1
+
+        # TODO/FIXME - Need to add logic to parse the topology to determine
+        #     egress ports being used on the switch.
+
+        replicas = [
```
Comment (spisarski) [inline trans_sec/controller/aggregate_controller.py:86]: Please add ticket so we don't forget to fix this

File: trans_sec/controller/core_controller.py [hunk 9]
```diff
@@ -123,19 +124,55 @@ def make_north_rules(self, sw, sw_info, north_link):
...
+                north_device.get('ip'), sw.name)
+
+    def set_multicast_group(self, sw, sw_info):
+        mc_group_id = 1
+
+        # TODO/FIXME - Need to add logic to parse the topology to determine
+        #     egress ports being used on the switch.
+
+        replicas = [
+            {'egress_port': '1', 'instance': '1'},
```
Comment (spisarski) [inline trans_sec/controller/core_controller.py:138]: Please add ticket so we don't forget to fix this

**Review by spisarski — CHANGES_REQUESTED** [review]
Looks pretty good. Please address the noted changes.

**Review by spisarski — CHANGES_REQUESTED** [review]
Also don't forget to add a task to the tps board for fixing the 'remote' testing.

**Comment by Sanjhana-Jayagopal** [comment]
Run Tests

File: p4/include/tps_headers.p4 [hunk 10]
```diff
@@ -189,15 +203,16 @@ struct headers {
...
-struct fwd_meta_t {
-    bit<32> l2ptr;
-    bit<24> out_bd;
+struct mac_learn_digest {
+    bit<32> src_IP;
+    bit<48> src_mac;
+    bit<9> ingress_port;
 }
 
 struct metadata {
```
Comment (spisarski) [inline p4/include/tps_headers.p4:217]: please remove this structure as it does not appear required any longer.

File: p4/include/tps_egress.p4 [hunk 11]
```diff
@@ -23,8 +23,16 @@ control TpsEgress(inout headers hdr,
                   inout metadata meta,
                   inout standard_metadata_t standard_metadata) {
 
-    apply {
+    action drop() {
+            mark_to_drop(standard_metadata);
```
Comment (spisarski) [inline p4/include/tps_egress.p4:27]: minor nit, indent 4 spaces from curly brackets

**Review by spisarski — APPROVED** [review]
All is good; however, I did add two minor comments.
Please check those out but I don't feel the need to review again :)

**Pull request merged**